Intro scene

Ohda opened the pull request:
**UPDATED 14/10** 
WIP Tag removed has i can't find obvious bugs.
I think i take care of all the comments excepts when @TomMalbran ask to move some Managers to GameController. I love the idea, but i don't find solutions for this without braking the game.
It can be a subject for another PR.
Just added the random background loading (but still 1 background for now)
*******

This is a basic intro scene with a main menu.
- Add a new GameController that will initialize some setup and persist between sceen.
- Add a new SceneController for loading, changing scenes.

Both scenes are able to be launched first. The new scene is in a _SCENES folder.
The Background is from @jaselito Thanks !

There is only one background but if there is more images in the Background folder the game will load a random image.

Known bug :
- There is a bug (also on master) when you open the settings screen without the fpscounter. ( #1526 )

![1](https://cloud.githubusercontent.com/assets/3942712/19215565/12022d06-8da2-11e6-8885-ae2455220899.png)
![2](https://cloud.githubusercontent.com/assets/3942712/19215566/120a464e-8da2-11e6-8787-b1961c69b695.png)
![3](https://cloud.githubusercontent.com/assets/3942712/19215567/120dc36e-8da2-11e6-9621-10c1537a74fc.png)

TomMalbran commented on Assets/Scripts/Models/Area/World.cs:
```diff
@@ -23,7 +23,7 @@ public class World : IXmlSerializable
     // TODO: Should this be also saved with the world data?
     // If so - beginner task!
-    public readonly string GameVersion = "Someone_will_come_up_with_a_proper_naming_scheme_later";
+    public static readonly string GameVersion = "Someone_will_come_up_with_a_proper_naming_scheme_later";
```
It seems so weird that the World store the GameVersion... it should be in Game.

TomMalbran commented on Assets/Scripts/Models/Events/TimeManager.cs:
```diff
@@ -120,13 +120,13 @@ public void Update(float time)
 
         // Systems that update every frame while unpaused.
-        if (WorldController.Instance.IsPaused == false)
+        if (IsPaused == false)
```
Notice that ```IsPaused == false``` is not always the same as ```GameController.Instance.IsPaused == false```, since ```GameController.Instance.IsPaused``` also checks for ```GameController.Instance.IsModal```.

TomMalbran commented on Assets/Scripts/UI/MainMenu.cs:
```diff
@@ -0,0 +1,49 @@
+using UnityEngine;
+using UnityEngine.UI;
+
+public class MainMenu : MonoBehaviour
```
We are already calling MainMenu to the bottom Menu in the game... should this be the MainScene?

TomMalbran commented on Assets/Scripts/Controllers/WorldController.cs:
```diff
@@ -54,7 +51,7 @@ public bool IsPaused
         get
         {
-            return TimeManager.Instance.IsPaused || IsModal;
+            return TimeManager.Instance.IsPaused || GameController.Instance.IsModal;
```
At this point it might make more sense to move this getter to GameController.

TomMalbran commented on Assets/Scripts/Controllers/GameController.cs:
```diff
@@ -0,0 +1,76 @@
+
+public class GameController : MonoBehaviour
+{
+    public KeyboardManager KeyboardManager;
```
You probably don't even need this. Unless we start removing some static classes.

jaselito commented:
The intro needs a little life... Here's something I made in Blender to use as a backdrop, at least for the time being (if you'd like, that is).  Sadly I couldn't find any free hi-res textures for mars, hence the distance from the planet.

![backdrop](https://cloud.githubusercontent.com/assets/21699103/18939684/9bca74d6-85c7-11e6-8f44-cdeb0ddeedd3.png)

TomMalbran commented:
@jaselito I like it.

Yes the intro page needs an image and a title too. We can go with a really simple logo for now, but it will be nice if it said Project Porcupine in the into (for now, as the name might change).

jaselito commented:
Project Porcupine should suffice for now; we just need to find a font everyone likes... :grin:

TomMalbran commented:
Yes. We could even use the font in all of quill18's project porcupine thumbnail videos until we get a proper logo.

UnknownSoldier1 commented:
@jaselito Is it possible to add stars and maybe some distant galaxies to your backdrop? That would make it perfect IMO.

jaselito commented:
Added some stars as requested; galaxies would be IMO overkill.
I personally prefer the no stars version :P 

![backdrop_stars](https://cloud.githubusercontent.com/assets/21699103/18965945/e4383726-8644-11e6-8ee5-ee8966c627ab.png)

dusho commented:
maybe try to put it through some cartoon and pixelize filters? so it's in sync with look of game?

jaselito commented:
Here a pixelated version (click to embiggen):

![backdrop_stars pixelated](https://cloud.githubusercontent.com/assets/21699103/18970203/f8c4bec4-8655-11e6-9a2d-f0903d7e4fc3.png)

I might give a try to adding asteroids, but those are much harder to pull off and I don't have any good free textures to use.

UnknownSoldier1 commented:
TBH, the non pixelated version is a lot better IMO.

Tranberry commented:
@jaselito I trust the texture was either Public Domain/C00 or any valid license that is compatible with PP? Better to ask the stupid questions than to regret it later.

TomMalbran commented:
I also prefer the non-pixalated version with the starts. We don't have to give the main screen a fully pixel look.

Tranberry commented:
![marshuh](https://cloud.githubusercontent.com/assets/16529/18975254/1e2eac26-86a9-11e6-8828-2b22bcb9e067.png)

Made a pixel version of your composition @jaselito not happy with the result.

jaselito commented:
@Tranberry I got the textures from [this site](http://planetpixelemporium.com/mars.html).  The author says the 1K versions (which I used) are free.  No additional info though. I think we should be safe; if not, oh well, it was a nice Blender exercise :stuck_out_tongue:
I actually think your pixel planet is looking quite good... maybe needs to be bigger and the shadow part should fade to black so it will mix nicely with a black/starry background.

jaselito commented:
@dusho I tried some asteroids with procedural textures:

![asteroids](https://cloud.githubusercontent.com/assets/21699103/18976984/252b5d76-867c-11e6-8d89-a933b502666a.png)

I know asteroids makes more sense for this game but I still like the planet better...

mikejbrown commented:
Actual space agency images are usually public domain. For example, here is [NASA's policy](https://www.nasa.gov/multimedia/guidelines/index.html). Basically you are free to use everything except logos for non-commercial purposes & are free to use most things even for commercial purposes as long as you don't a) imply NASA's endorsement and b) use the image(s) in a misleading way and c) do not show astronauts or employees without permission. Policies for other space agencies are probably similar.

So you may want to trawl their [image gallery](https://www.nasa.gov/multimedia/imagegallery/index.html) if you want a more realistic look.

jaselito commented:
Yes, I knew about NASA's policies and actually looked through that site.  Problem is, optimally we need textures fit to wrap around a 3d model, at least a diffuse and a normal map.  Just pictures of celestial objects aren't usually enough.

Tranberry commented:
> _Rule one never use true black as you probably want to add something darker later, rule two never use true white as you probably want to use something lighter later._ - Some digital artist dude

Tranberry commented:
I'd try to stay away from real world textures too, not for inspiration but for direct use. Are we even in the 'real' world?

dusho commented:
maybe water paint filter and some blur?
![image](https://cloud.githubusercontent.com/assets/12616499/18993598/40f031dc-8724-11e6-9c44-17e599f16da6.png)

Ohda commented:
Hello ! I've got à New home and i'am waiting internet. Can't wait to come Back on this !

jaselito commented:
@Tranberry I agree that that rule makes sense in a general way, but not in this case; there isn't any blacker than the blackness of space :grin:   Of course, I only say fade to black if we intend to put the planet in a black background... other background colors would be a different story.

@dusho Me likes!

@Ohda Ah, no internet, I feel you pain... can't even remember how pre-internet times were...

mikejbrown commented:
This pull request has merge conflicts. Please resolve the conflicts and remove [Conflicts] from the title.

Ohda commented:
To give you a quick status, i'am still waiting my new internet connexion, but i'am trying to work on this offline. 
I've done a SceneManagerProjectProcupine for all the loadScene, QuitGame method.
I have use the DB_PromptOrInfo.prefab to do a loading message.
Added a background picture (iam not sure witch one to take).
I have added a Settings and a load game button and i will start having a look on implementing the functionnalities.
If someone can have a look for a basic first Title it can be great !

Tranberry commented:
>  Added a background picture (iam not sure witch one to take).

I'd have an animated one if at all possible and preferable driven by code, nothing fancy but some blinking stars and maybe a ship fly by. I'm just dreaming now..

Ohda commented:
![1](https://cloud.githubusercontent.com/assets/3942712/19141579/ea3d49de-8b96-11e6-9871-7ffdfb400d8a.png)
![2](https://cloud.githubusercontent.com/assets/3942712/19141578/ea3cec28-8b96-11e6-8c70-8469278d0486.png)
![3](https://cloud.githubusercontent.com/assets/3942712/19141580/ea47df66-8b96-11e6-9cc4-f08374004ab3.png)

dusho commented:
I would go for asteroids.. as I haven't seen such a nice red planet in the game.. (would be a false advertisement, I guess :) )

UnknownSoldier1 commented:
Maybe add this nice red planet later into the game and then it's no longer false advertising ^^

TomMalbran commented:
I like the planet one better, and we can totally add it to the game at some point.

The title could use some work, but it works for now.

Good job.

jaselito commented:
I like 'em both :D  
Maybe make the text bluish and make sure text and buttons don't get too small.

jaselito commented:
Doing some experimenting...

![planet_asteroids_stars](https://cloud.githubusercontent.com/assets/21699103/19168568/4a0777c4-8bd6-11e6-992b-1ee1fab15fc9.png)

Tranberry commented:
I like it but the composition seems unbalanced (like a unoptimal photo of a beautiful vista)

jaselito commented:
@Tranberry I can move asteroids around... any idea for balancing?

Tranberry commented:
It might just be the missing menu, I'm not sure.

jaselito commented:
Well, yes, I'm deliberately leaving the right side empty for menu and title...

mikejbrown commented:
Maybe it's too many asteroids?

jaselito commented:
Ok, moved things around a little bit: 

![planet_asteroids_stars 2](https://cloud.githubusercontent.com/assets/21699103/19175479/8b13d1d8-8bfc-11e6-890f-538214aa8a91.png)

bjubes commented:
maybe a dumb idea, but if we can't agree on one title image we could have a bunch that cycle (kinda like win 10 lock screen)

TomMalbran commented:
I like the last image.

We can later have a random image each time the game loads.

Tranberry commented:
Later it should be animated ;P 
It looks much better imo now Jaselito

jaselito commented:
I'm also happy with the end result.. thanks for all your input.  Later we can think of ways to animate it, but I think they should be very subtle (the blinking stars @Tranberry  mentioned, asteroids slowly drifting...)

NogginBops commented:
How about we render the asteroids as a planetary ring. Would explain the asteroids in the game. :)

jaselito commented:
After too much tweaking for my own sanity this is the best I could do to make it seem like we're dealing with an asteroid ring... not sure I can make it work :confused:

![planet_asteroids_ring_stars](https://cloud.githubusercontent.com/assets/21699103/19209342/08316d02-8cce-11e6-832b-416a9ad95a39.png)

Ohda commented:
Remove the WIP tag. Update the original post.

TomMalbran commented on Assets/Scripts/Controllers/WorldController.cs:
```diff
@@ -119,15 +98,14 @@ public void Start()
         // Register inputs actions
-        keyboardManager.RegisterInputAction("Pause", KeyboardMappedInputType.KeyUp, () => { IsPaused = !IsPaused; });
-        keyboardManager.RegisterInputAction("DevMode", KeyboardMappedInputType.KeyDown, ChangeDevMode);
+        KeyboardManager.Instance.RegisterInputAction("Pause", KeyboardMappedInputType.KeyUp, () => { GameController.Instance.IsPaused = !GameController.Instance.IsPaused; });
```
We should probably move this line to the GameController. Makes more sense to have it there now.

TomMalbran commented on Assets/Scripts/UI/InGameUI/MenuController.cs:
```diff
@@ -0,0 +1,147 @@
+
+    private Button CreateButton(string text)
+    {
+        GameObject buttonQuestGameObject = (GameObject)Instantiate(Resources.Load("UI/MenuButton"), this.gameObject.transform);
```
You should change the name of the variable.
Is it a good idea to start having UI utilities for functions like this one?

TomMalbran commented on Assets/Scripts/Utilities/SceneManagerProjectPorcupine.cs:
```diff
@@ -0,0 +1,80 @@
+#endregion
+using UnityEngine.SceneManagement;
+
+public class SceneManagerProjectPorcupine
```
I would prefer if `ProjectPorcupine` was part of the namespace, instead of part of the name.

TomMalbran replied:
I also think that it is more of a Controller and should go in the Controllers folder.

TomMalbran commented on Assets/Scripts/Models/InputOutput/ModsManager.cs:
```diff
@@ -14,13 +14,26 @@ public class ModsManager
-        DirectoryInfo modsDir = new DirectoryInfo(modsPath);
-        mods = modsDir.GetDirectories();
+        mods = GetModsFiles();
+        if (SceneManagerProjectPorcupine.Instance.GetCurrentScene() == SceneManagerProjectPorcupine.MainMenuSceneName)
```
It feels like this could be a method `SceneManagerProjectPorcupine.isAtMainMenu()` or similar.

TomMalbran commented on Assets/Scripts/Models/Events/TimeManager.cs:
```diff
@@ -120,13 +120,13 @@ public void Update(float time)
 
         // Systems that update every frame while unpaused.
-        if (WorldController.Instance.IsPaused == false)
+        if (IsPaused == false && GameController.Instance.IsModal == false)
```
This is the same as `GameController.Instance.IsPaused == false`.

TomMalbran commented on Assets/Scripts/Controllers/MainMenuController.cs:
```diff
@@ -0,0 +1,63 @@
+
+        new SpriteManager();
+        new AudioManager();
+        modsManager = new ModsManager();
```
Should we load these here or in the GameController? I think that it might be clear to just leave this to deal with the MainMenu and not with loading stuff. By "these" I mean PrototypeManager, SpriteManager, AudioManager and ModsManager.

bjubes commented:
after switching language and then applying/saving, the buttons do not switch text. They only update to the new language after stopping and restarting the game

Ohda commented:
I've got an issue coming from merging #1490. Settings doesn't apply/save correctly. 
WIP tag readded.

Ohda commented:
Update the orginal post.

bjubes commented:
now that this is in we should probably have the in game quit button bring the player to the main menu, but thats another PR.

bjubes commented:
Sorry, didn't see that this breaks loading from the _SCENE_, but not from main menu. I'm reverting this pull, so don't delete the branch
